torch/distributed/algorithms/_checkpoint/checkpoint_wrapper.py

```diff
@@ -155,10 +92,107 @@ def _pre_load_state_dict_hook(
         _replace_by_prefix(state_dict, prefix, prefix + f"{_CHECKPOINT_PREFIX}.")
 
 
+class OffloadWrapper(ActivationWrapper):
+    def __init__(self, mod):
+        super().__init__(mod)
+
+    def forward(self, *args, **kwargs):
+        with save_on_cpu(pin_memory=True):
```

Comment:
Tangentially related: 

> You should not over-allocate pinned memory. Doing so can reduce overall system performance because it reduces the amount of physical memory available to the operating system and other programs. How much is too much is difficult to tell in advance, so as with all optimizations, test your applications and the systems they run on for optimal performance parameters.

https://developer.nvidia.com/blog/how-optimize-data-transfers-cuda-cc/

For activation offload, we operate under the assumption that we have unlimited memory to pin. I think it is worthwhile to investigate the robustness of this assumption.

Without any offloading, the total activation size is upper bounded by the GPU memory size, but with offloading, that can be increased significantly (e.g. <= the number of activations times the GPU memory size). For some workloads that activation offloading targets (e.g. large batch size and activations; small model), I imagine that the total activation size can be larger than the GPU memory size. This is exacerbated if CPU offloading is simultaneously enabled.

That being said, if the CPU memory size is sufficiently large and not used for much else during program execution, then pinning significant amounts of CPU memory may not be a problem. I think we should simply try to get some better intuition here.

Reply:
This makes sense. I filed an issue: https://github.com/pytorch/pytorch/issues/86097. In the medium term if we do determine we shouldn't always pin memory, we might make this flag configurable or in the long-term possibly something that can be autotuned.